### PR TITLE
[webgpu] Add batchNormal kernel

### LINKED ISSUE #2358: [webgpu] Add batchNorm kernel.
Used in handtracking.

### PR DESCRIPTION
**Pull request by NALLEIN**
Fix #2358 
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/2652)

### CODE REVIEW

File: tfjs-backend-webgpu/src/setup_test.ts [hunk 1]
```diff
@@ -331,7 +331,8 @@ const TEST_FILTERS: TestFilter[] = [
                           // reason
       'MultipleBoxes-DifferentBoxes',  // TimeOut
     ]
-  }
+  },
+  {include: 'deprecated batchNormalization', excludes: []}
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/setup_test.ts]: Do the non deprecated batchnorm tests not work?

File: tfjs-backend-webgpu/src/backend_webgpu.ts [hunk 2]
```diff
@@ -549,6 +550,30 @@ export class WebGPUBackend extends KernelBackend {
                 input.size < sizeThreshold);
   }
 
+  batchNormalization(
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/backend_webgpu.ts]: We are moving towards a fully modular setup - would you mind extracting batchNormalization into its own kernel file, similar to the `Square` op? WebGL and CPU backends have not yet done this, but they will.

File: tfjs-backend-webgpu/src/kernels/BatchNormal.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {KernelConfig, FusedBatchNorm, FusedBatchNormInputs,FusedBatchNormAttrs, Tensor} from '@tensorflow/tfjs-core';
```
Comment (axinging) [inline tfjs-backend-webgpu/src/kernels/BatchNormal.ts:18]: Missing FusedBatchNormInputs and FusedBatchNormAttrs? These shoube be defined in the tfjs-cores.

Comment (NALLEIN) [reply tfjs-backend-webgpu/src/kernels/BatchNormal.ts:18]: It was defined in https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/kernel_names.ts#L36 and it has the same inputs as batchNorm so I reuse it

Comment (annxingyuan) [reply tfjs-backend-webgpu/src/kernels/BatchNormal.ts:18]: We are trying to match TensorFlow for kernel names as much as possible - the batchNormalization kernel is called `FusedBatchNorm` in TensorFlow.

File: tfjs-backend-webgpu/src/kernels/BatchNormal.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import {WebGPUBackend} from '../backend_webgpu';
+import {BatchNormProgram} from './batchnorm_webgpu';
+
+export const batchNromConfig : KernelConfig = {
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/kernels/BatchNormal.ts:22]: Could you rename this file to be FusedBatchNorm.ts? See our WASM kernel for reference. This matches TensorFlow's kernel naming scheme.

File: tfjs-backend-webgpu/src/kernels/BatchNormal.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            offsetShape = offset.shape;
+            batchNormInputs.push(offset as Tensor);
+        }
+        let scaleShape = null;
+        if(scale != null) {
+            scaleShape = scale.shape;
+            batchNormInputs.push(scale as Tensor);
+        }
+        const program = new BatchNormProgram(
+            x.shape,mean.shape,variance.shape,offsetShape,scaleShape,
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/kernels/BatchNormal.ts:41]: Could you install our code formatter clang? Otherwise whitespace diffs will appear in later commits :)

Details for installation here: https://github.com/tensorflow/tfjs/blob/master/DEVELOPMENT.md

File: tfjs-backend-webgpu/src/kernels/Rsqrt.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import {RSQRT, UnaryOpProgram} from './unary_op_webgpu';
+
+export const rsqrtConfig: KernelConfig = {
+  kernelName: Rsqrt,
+  backendName: 'webgpu',
+  kernelFunc: ({inputs, backend}) => {
+    const {x} = inputs as RsqrtInputs;
+    const webGPUBackend = backend as WebGPUBackend;
+    const program = new UnaryOpProgram(x.shape, RSQRT);
+    return webGPUBackend.compileAndRun(program, [x as Tensor]);
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/kernels/Rsqrt.ts:29]: Is this cast necessary?

**Comment by NALLEIN** [comment]
> Left a few more comments!

I fixed some problems, please review.

File: tfjs-backend-webgpu/src/kernels/batchnorm_webgpu.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    switch (dim) {
+      case 2:
+        this.dispatchLayout = {x: [1], y: [0], z: []};
+        setOutput = 'setOutput(coords[0], coords[1], value);';
+        break;
+      case 3:
+        this.dispatchLayout = {x: [1, 2], y: [0], z: []};
+        setOutput = 'setOutput(coords[0], coords[1], coords[2], value);';
+        break;
+      case 4:
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/kernels/batchnorm_webgpu.ts:56]: Looks like case 4 and default are the same - can you use fallthrough syntax to avoid duplication?

File: tfjs-backend-webgpu/src/kernels/batchnorm_webgpu.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        this.dispatchLayout = {x: [1, 2], y: [0], z: [3]};
+        setOutput =
+            'setOutput(coords[0], coords[1], coords[2], coords[3],value);';
+        break;
+      default:
+        this.dispatchLayout = {x: [1, 2], y: [0], z: [3]};
+        setOutput =
+            'setOutput(coords[0], coords[1], coords[2], coords[3],value);';
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/kernels/batchnorm_webgpu.ts:64]: Can we use if/else rather than switch? Also we know that the output and x are 4D - can we avoid this logic altogether? We often reshape tensors before passing them to kernels to make it possible to simplify shader code like this.

**Comment by annxingyuan** [comment]
@NALLEIN A few more tiny comments!

File: tfjs-backend-webgpu/src/kernels/batchnorm_webgpu.ts (hunk 8 above)
Comment (NALLEIN) [reply tfjs-backend-webgpu/src/kernels/batchnorm_webgpu.ts]: I fix the syntax. I think this logic needs to be preserved. The x and output may be 2D, 3D, or 4D because ops is modified in #2996 ; batchnorm_2d, batchnorm_3d, and batchnorm_4d will call this kernel.

File: tfjs-backend-webgpu/src/kernels/FusedBatchNorm.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+ * =============================================================================
+ */
+
+import {FusedBatchNorm, FusedBatchNormAttrs, FusedBatchNormInputs, KernelConfig, Tensor} from '@tensorflow/tfjs-core';
+
+import {WebGPUBackend} from '../backend_webgpu';
+
+import {BatchNormProgram} from './batchnorm_webgpu';
+
+export const fusedBatchNromConfig: KernelConfig = {
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/kernels/FusedBatchNorm.ts:24]: typo: fusedBatchNormConfig

File: tfjs-backend-webgpu/src/register_all_kernels.ts [hunk 10]
```diff
@@ -16,12 +16,13 @@
...
 import {KernelConfig, registerKernel} from '@tensorflow/tfjs-core';
 
+import {fusedBatchNromConfig} from './kernels/FusedBatchNorm';
 import {squareConfig} from './kernels/Square';
 import {squaredDifferenceConfig} from './kernels/SquaredDifference';
 
 // List all kernel configs here
 const kernelConfigs: KernelConfig[] =
-    [squareConfig, squaredDifferenceConfig];
+    [squareConfig, squaredDifferenceConfig, fusedBatchNromConfig];
```
Comment (annxingyuan) [inline tfjs-backend-webgpu/src/register_all_kernels.ts:25]: typo: fusedBatchNormConfig

**Comment by annxingyuan** [comment]
Oh also if you wouldn't mind merging in master - we might have a few overlapping changes!

**Comment by NALLEIN** [comment]
> Sorry for the delay on this... just two more typos and we'll be all set!

Sorry，I've fixed the name issue.
